File: C++/386. Lexicographical Numbers.cpp
class Solution {
    vector<int> ans;
    int num;
public:
    void helper(int n){
        if(n > num) return;
        ans.push_back(n);
        for(int i = 0; i<=9; i++){
            helper(n*10 + i);
        }
    }
    vector<int> lexicalOrder(int n) {
        ans = vector<int>(0);
        num = n;
        for(int i = 1; i<=9; i++) helper(i);
        return ans;
    }
};

File: JAVA/386. Lexicographical Numbers.java
class Solution {
    public List<Integer> lexicalOrder(int n) {
        List<Integer> ans = new ArrayList<>(); 
        for(int i = 1; i <= 9; i++) {
            generate(ans, i, n);  
        }        
        return ans;  
    }
    void generate(List<Integer> ans, int curr, int limit) {
        if (curr > limit) {
            return;
        }
        ans.add(curr);
        for (int i = 0; i <= 9; i++) {
            int newCurr = (curr * 10) + i;  
            generate(ans, newCurr, limit); 
        }
    }
}
